### DISCUSSION TITLE
Any way to double-check that utf-validate and bufferutil are being used?

### DISCUSSION THREAD

**Discussion by trekze** [socketio/socket.io#4134] [Q&A] [answered]
I'm seeing 3x the CPU load between two servers with the same number of users. These servers score identically on single-threaded benchmarks  (One is Intel the other is AMD). They both have bufferutil and utf-validate installed. Is there a way for me to check if they are both using the libraries, and what could be anothe reason for the massive performance gap between them?

Thank you.

**Comment by trekze** [comment]
I ran some profiling and it seems `writev` is the function that takes 3x as long on the Intel server, despite equal single-thread performance bencharks.

**Comment by darrachequesne** [comment, marked as answer]
I think you can check by going into the `node_modules/bufferutil` directory, and then run:

```js
$ node
> require('node-gyp-build')(".")
// should print "{ mask: [Function], unmask: [Function] }"
```

Reference: https://github.com/websockets/bufferutil/blob/330205e62cf079371438c2ddc2082931a703e6a1/index.js#L4

If it throws, that means it's the JS implementation that is used instead of the native one.

**Comment by trekze** [comment]
Thank you. Seems like it's being used. 97% of the profiling time is spent in C++ code so I guess everything is working as it should. Guess I'm just running into the limits of putting stuff on the wire with socketio. 

You probably can't answer this, but do you think it's reasonable for the following workload to consume 50% cpu of a single core, on modern hardware?

- OUT: 20 messages per second to 200 clients (i.e. 4000 messages per second @ 300 bytes each)
- IN: 20 message per second from 200 clients (i.e. 4000 messages per second @ 10 bytes each)
- TOTAL: 8000 messages per second.

Very very roughly speaking.

**Reply by darrachequesne** [reply]
That's indeed a bit high. Did you check with `eiows` (see [here](https://socket.io/docs/v4/server-installation/#other-websocket-server-implementations))?

Besides, we are currently working on an implementation based on `uWebSocket.js`, a high performing HTTP/WebSocket server: https://github.com/socketio/engine.io/releases/tag/6.1.0
